File: cpp/src/network_analysis.cpp
#include "metro/network_analysis.hpp"
#include "metro/graph.hpp"
#include "metro/station.hpp"

#include <queue>
#include <stack>
#include <unordered_set>
#include <algorithm>

namespace metro::analysis {

// ---------------------------------------------------------------------------
// affected_area — BFS K-order neighbor analysis
// ---------------------------------------------------------------------------

std::vector<std::string> affected_area(const Graph& graph,
                                        const StationManager& mgr,
                                        const std::string& station_id,
                                        int max_depth) {
    std::vector<std::string> result;
    std::unordered_set<std::string> visited;
    std::queue<std::pair<std::string, int>> q;  // (node_id, depth)

    q.push({station_id, 0});
    visited.insert(station_id);

    const auto& adj = graph.adj();

    while (!q.empty()) {
        auto [cur, depth] = q.front(); q.pop();

        if (depth >= max_depth) continue;

        auto it = adj.find(cur);
        if (it == adj.end()) continue;

        for (const auto& edge : it->second) {
            const auto& next_id = edge.to_id;
            if (visited.count(next_id)) continue;
            visited.insert(next_id);

            // Only include open stations in affected area
            const Station* station = mgr.get(next_id);
            if (station && station->is_open()) {
                result.push_back(next_id);
            }

            q.push({next_id, depth + 1});
        }
    }

    // Sort for deterministic output
    std::sort(result.begin(), result.end());
    return result;
}

// ---------------------------------------------------------------------------
// count_components — DFS connected components
// ---------------------------------------------------------------------------

std::vector<std::vector<std::string>> count_components(
    const Graph& graph,
    const StationManager& mgr) {

    std::unordered_set<std::string> visited;
    std::vector<std::vector<std::string>> components;

    const auto& adj = graph.adj();

    // Iterative DFS over all open stations
    for (const auto& [id, s] : mgr.stations()) {
        if (!s.is_open()) continue;
        if (visited.count(id)) continue;

        std::vector<std::string> component;
        std::stack<std::string> stack;
        stack.push(id);

        while (!stack.empty()) {
            std::string node = stack.top(); stack.pop();
            if (visited.count(node)) continue;
            visited.insert(node);
            component.push_back(node);

            auto it = adj.find(node);
            if (it == adj.end()) continue;

            for (const auto& edge : it->second) {
                const auto& next_id = edge.to_id;
                if (visited.count(next_id)) continue;

                const Station* station = mgr.get(next_id);
                if (!station || !station->is_open()) {
                    visited.insert(next_id);  // mark as visited to skip
                    continue;
                }

                stack.push(next_id);
            }
        }

        if (!component.empty()) {
            components.push_back(std::move(component));
        }
    }

    // Sort by size descending (largest component first)
    std::sort(components.begin(), components.end(),
              [](const auto& a, const auto& b) {
                  return a.size() > b.size();
              });

    return components;
}

} // namespace metro::analysis

File: src/network_analysis.py
"""
network_analysis.py — 网络分析功能
====================================

提供：
  - affected_area()     受关闭站点影响分析（BFS K 阶邻居）
  - count_components()  全网连通分量检测（DFS）
"""

from __future__ import annotations

from collections import deque
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graph import Graph
    from station import StationManager


def affected_area(graph: Graph,
                  station_mgr: StationManager,
                   station_id: str,
                   max_depth: int = 2) -> list[str]:
    """BFS 查找受关闭站点影响的 K 阶邻居（不含关闭站本身）。

    参数:
      station_id: 被关闭的站点 ID
      max_depth:  分析深度（默认 2 阶）

    返回: [affected_station_id, ...]（按 BFS 顺序）
    """
    if max_depth <= 0:
        return []

    visited: set[str] = set()
    queue: deque[tuple[str, int]] = deque()
    affected: list[str] = []

    station = station_mgr.get(station_id)
    if station is None:
        return []

    queue.append((station_id, 0))
    visited.add(station_id)

    while queue:
        cur, depth = queue.popleft()
        if depth >= max_depth:
            continue

        for edge in graph._adj.get(cur, []):
            neighbor = edge.to_id
            if neighbor not in visited:
                visited.add(neighbor)
                # 只算其他线路上同名站点的受影响
                ns = station_mgr.get(neighbor)
                if ns and ns.is_open:
                    affected.append(neighbor)
                    queue.append((neighbor, depth + 1))

    return affected


def count_components(graph: Graph,
                     station_mgr: StationManager) -> list[list[str]]:
    """DFS 求连通分量（仅考虑开启的站点）。

    返回: [[component_station_ids], ...]，按分量大小降序
    """
    visited: set[str] = set()
    components: list[list[str]] = []

    def dfs(node: str, comp: list[str]) -> None:
        visited.add(node)
        comp.append(node)
        for edge in graph._adj.get(node, []):
            neighbor = edge.to_id
            if neighbor not in visited:
                ns = station_mgr.get(neighbor)
                if ns is None or not ns.is_open:
                    continue
                dfs(neighbor, comp)

    all_ids = graph.all_ids()
    for sid in all_ids:
        if sid in visited:
            continue
        s = station_mgr.get(sid)
        if s is None or not s.is_open:
            visited.add(sid)
            continue
        comp: list[str] = []
        dfs(sid, comp)
        components.append(comp)

    components.sort(key=len, reverse=True)
    return components


# ---------------------------------------------------------------------------
# 自测
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    from graph import Graph
    from station import StationManager

    g = Graph()
    g.load()
    mgr = StationManager()
    mgr.load()

    # 关闭一个站
    for s in mgr.find_by_name("漕宝路"):
        if "1号线" in s.line:
            mgr.close_station(s.station_id)
            print(f"关闭 {s}")

    # 受影响分析
    print(f"\n受影响站点（1阶）:")
    for s in mgr.find_by_name("漕宝路"):
        if not s.is_open:
            affected = affected_area(g, mgr, s.station_id, max_depth=1)
            for aid in affected:
                ns = mgr.get(aid)
                if ns:
                    print(f"  {ns}")

    # 连通分量
    print(f"\n连通分量:")
    comps = count_components(g, mgr)
    print(f"总分量数: {len(comps)}")
    for i, comp in enumerate(comps[:3]):
        print(f"  分量 {i+1}: {len(comp)} 个节点")

    # 恢复
    mgr.restore_initial()